org-roam.el

```diff
@@ -398,17 +398,26 @@ which takes as its argument an alist of path-completions.  See
 ;;;; org-roam-find-index
 (defcustom org-roam-index-file nil
   "Path to the Org-roam index file.
-The path can either be relative (to `org-roam-directory') or
-absolute.
-If it is not defined, the index is assumed to be a note in
-`org-roam-directory' whose title is 'Index'.")
+The path can be a string or a function.  If it is a string, it
+should be the path (absolute or relative to `org-roam-directory')
+to the index file.  If it is is a function, the function should
+return the path to the index file.  Otherwise, the index is
+assumed to be a note in `org-roam-directory' whose title is
+'Index'."
+  :type '(choice
+          (string :tag "Path to index" "%s")
+          (function :tag "Function to generate the path"))
+  :group 'org-roam)
 
-(defun org-roam--find-index ()
+(defun org-roam--get-index-path ()
   "Return the path to the index in `org-roam-directory'.
 The path to the index can be defined in `org-roam-index-file'.
 Otherwise, it is assumed to be a note in `org-roam-directory'
 whose title is 'Index'."
-  (let ((path org-roam-index-file))
+  (let ((path (--> org-roam-index-file
+                   (if (functionp it)
+                       (funcall it)
+                     it))))
```

Comment:
Should probably verify that `org-roam-index-file` is a function or a string explicitly. Throw an error in cases where it is not.

Reply:
```emacs-lisp

  (pcase org-roam-index-file
    ((pred functionp) (funcall org-roam-index-file))
    ((pred stringp)   org-roam-index-file)
    (wrong-type (signal 'wrong-type-argument `((functionp stringp) ,wrong-type))))
```

Reply:
…Wow.  I didn't know you could do [all this with `pcase`](http://newartisans.com/2016/01/pattern-matching-with-pcase/)!  Colour me impressed, and thanks for the discovery!

Reply:
Not only that, but you've taught me how to use `signal`.  I'm completely mindblown by the amount of stuff I've discovered in just 4 lines.  Really, thanks!

Reply:
Glad to help! Pattern matching is a powerful feature.
At some point we may want to issue our own types of errors, but for now using signal with the standard errors is better than nothing at all.